File: src/main/java/com/example/dq/repository/DataSourceRepository.java
package com.example.dq.repository;

import com.example.dq.model.DataSourceConfig;
import com.example.dq.model.DbType;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;
import org.springframework.stereotype.Repository;

import java.sql.PreparedStatement;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Repository
public class DataSourceRepository {

    private final JdbcTemplate jdbc;

    public DataSourceRepository(JdbcTemplate jdbc) {
        this.jdbc = jdbc;
    }

    /** password_enc 已加密存储,读取后由调用方解密 */
    private final RowMapper<DataSourceConfig> mapper = (rs, i) -> {
        DataSourceConfig c = new DataSourceConfig();
        c.setId(rs.getLong("id"));
        c.setName(rs.getString("name"));
        c.setDbType(DbType.valueOf(rs.getString("db_type")));
        c.setJdbcUrl(rs.getString("jdbc_url"));
        c.setUsername(rs.getString("username"));
        c.setPassword(rs.getString("password_enc"));
        long rt = rs.getLong("row_threshold");
        c.setRowThreshold(rs.wasNull() ? null : rt);
        long st = rs.getLong("size_threshold_bytes");
        c.setSizeThresholdBytes(rs.wasNull() ? null : st);
        return c;
    };

    public long insert(DataSourceConfig c) {
        KeyHolder kh = new GeneratedKeyHolder();
        jdbc.update(conn -> {
            PreparedStatement ps = conn.prepareStatement(
                    "INSERT INTO data_source(name, db_type, jdbc_url, username, password_enc, row_threshold, size_threshold_bytes) "
                            + "VALUES (?,?,?,?,?,?,?)", new String[]{"ID"});
            ps.setString(1, c.getName());
            ps.setString(2, c.getDbType().name());
            ps.setString(3, c.getJdbcUrl());
            ps.setString(4, c.getUsername());
            ps.setString(5, c.getPassword());
            if (c.getRowThreshold() != null) ps.setLong(6, c.getRowThreshold()); else ps.setNull(6, java.sql.Types.BIGINT);
            if (c.getSizeThresholdBytes() != null) ps.setLong(7, c.getSizeThresholdBytes()); else ps.setNull(7, java.sql.Types.BIGINT);
            return ps;
        }, kh);
        return Objects.requireNonNull(kh.getKey()).longValue();
    }

    public void update(DataSourceConfig c, boolean updatePassword) {
        String sql = "UPDATE data_source SET name=?, db_type=?, jdbc_url=?, username=?, row_threshold=?, size_threshold_bytes=?, updated_at=CURRENT_TIMESTAMP"
                + (updatePassword ? ", password_enc=?" : "") + " WHERE id=?";
        jdbc.update(conn -> {
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setString(1, c.getName());
            ps.setString(2, c.getDbType().name());
            ps.setString(3, c.getJdbcUrl());
            ps.setString(4, c.getUsername());
            if (c.getRowThreshold() != null) ps.setLong(5, c.getRowThreshold()); else ps.setNull(5, java.sql.Types.BIGINT);
            if (c.getSizeThresholdBytes() != null) ps.setLong(6, c.getSizeThresholdBytes()); else ps.setNull(6, java.sql.Types.BIGINT);
            int idx = 7;
            if (updatePassword) ps.setString(idx++, c.getPassword());
            ps.setLong(idx, c.getId());
            return ps;
        });
    }

    public void delete(long id) {
        jdbc.update("DELETE FROM data_source WHERE id=?", id);
    }

    public Optional<DataSourceConfig> findById(long id) {
        List<DataSourceConfig> list = jdbc.query("SELECT * FROM data_source WHERE id=?", mapper, id);
        return list.stream().findFirst();
    }

    public List<DataSourceConfig> findAll() {
        return jdbc.query("SELECT * FROM data_source ORDER BY id", mapper);
    }
}

File: desktop/src/main/kotlin/com/example/dq/repository/DataSourceRepository.kt
package com.example.dq.repository

import com.example.dq.model.DataSourceConfig
import com.example.dq.model.DbType
import java.sql.ResultSet
import java.sql.Types

class DataSourceRepository(private val jdbc: Jdbc) {

    /** password_enc 已加密存储,读取后由调用方解密 */
    private val mapper: (ResultSet) -> DataSourceConfig = { rs ->
        val c = DataSourceConfig()
        c.id = rs.getLong("id")
        c.name = rs.getString("name")
        c.dbType = DbType.valueOf(rs.getString("db_type"))
        c.jdbcUrl = rs.getString("jdbc_url")
        c.username = rs.getString("username")
        c.password = rs.getString("password_enc")
        val rt = rs.getLong("row_threshold")
        c.rowThreshold = if (rs.wasNull()) null else rt
        val st = rs.getLong("size_threshold_bytes")
        c.sizeThresholdBytes = if (rs.wasNull()) null else st
        c.dbMode = rs.getString("db_mode")
        c
    }

    fun insert(c: DataSourceConfig): Long =
        jdbc.insert(
            "INSERT INTO data_source(name, db_type, jdbc_url, username, password_enc, row_threshold, size_threshold_bytes, db_mode) " +
                    "VALUES (?,?,?,?,?,?,?,?)",
            c.name, c.dbType!!.name, c.jdbcUrl, c.username, c.password, c.rowThreshold, c.sizeThresholdBytes, c.dbMode)

    fun update(c: DataSourceConfig, updatePassword: Boolean) {
        val sql = "UPDATE data_source SET name=?, db_type=?, jdbc_url=?, username=?, row_threshold=?, size_threshold_bytes=?, db_mode=?, updated_at=CURRENT_TIMESTAMP" +
                (if (updatePassword) ", password_enc=?" else "") + " WHERE id=?"
        jdbc.withStatement({ conn -> conn.prepareStatement(sql) }) { ps ->
            ps.setString(1, c.name)
            ps.setString(2, c.dbType!!.name)
            ps.setString(3, c.jdbcUrl)
            ps.setString(4, c.username)
            val rt = c.rowThreshold
            if (rt != null) ps.setLong(5, rt) else ps.setNull(5, Types.BIGINT)
            val st = c.sizeThresholdBytes
            if (st != null) ps.setLong(6, st) else ps.setNull(6, Types.BIGINT)
            ps.setString(7, c.dbMode)
            var idx = 8
            if (updatePassword) ps.setString(idx++, c.password)
            ps.setLong(idx, c.id!!)
            ps.executeUpdate()
        }
    }

    fun delete(id: Long) {
        jdbc.update("DELETE FROM data_source WHERE id=?", id)
    }

    fun findById(id: Long): DataSourceConfig? =
        jdbc.queryOne("SELECT * FROM data_source WHERE id=?", id, mapper = mapper)

    fun findAll(): List<DataSourceConfig> =
        jdbc.query("SELECT * FROM data_source ORDER BY id", mapper = mapper)
}
